### ISSUE TITLE
[Enhancement]: Support CodePipeline stage level conditions

### ISSUE THREAD

**Issue by egoaty** [hashicorp/terraform-provider-aws#39284]
### Description

AWS recently announced stage level conditions to implement pipeline gates for CodePipeline.

Announcement: https://aws.amazon.com/about-aws/whats-new/2024/08/aws-codepipeline-stage-level-conditions-implement-pipeline-gates/

Documentation: https://docs.aws.amazon.com/codepipeline/latest/userguide/stage-conditions.html

### Affected Resource(s) and/or Data Source(s)

aws_codepipeline

### Potential Terraform Configuration

```terraform
resource "aws_codepipeline" "pipeline" {
  name = "my-pipeline"
  ...
  stage {
    name       = "Deploy"
    ...
    before_entry {
      condition {
        result = "FAIL"
        rule {
          name = "DeploymentWindow"
          rule_type_id = {
            category = "Rule"
            owner = "AWS"
            provider = "DeploymentWindow"
            version = "1"
          }
          configuration = {
            Cron = "* * 9-17 ? * MON-FRI *"
            TimeZone = "UTC"
          }
          ...
        }
      }
    }
  }
}
```

### References

https://aws.amazon.com/about-aws/whats-new/2024/08/aws-codepipeline-stage-level-conditions-implement-pipeline-gates/
https://docs.aws.amazon.com/codepipeline/latest/userguide/stage-conditions.html
https://awsapichanges.com/archive/changes/021533-codepipeline.html

### Would you like to implement a fix?

None

**Comment by gwrite** [comment]
Very useful this enhancement!!

**Comment by roger-watkins** [comment]
I am trying to implement this right now as i've been looking for an alternative to using a lambda function to skip a stage on a certain condition - in my case funnily enough i'm running a terraform pipeline and I want to skip the approval/deploy stage when there's no changes.

This enhancement would be very welcome.

**Comment by RealArv** [comment]
I agree, something in the lines of what OP has shown would be great. Currently I'm handling similar situations through buildspec by passing an env variable and doing a conditional check which will run a section only if the check passes, but it has its own limitations. Having this could and would be a better practice and solution.
